### PR TITLE
[ALLUXIO-2914] Improve setTtl command to take the time-to-live value with unit

### PR DESCRIPTION
**Pull request by maboelhassan**
https://alluxio.atlassian.net/browse/ALLUXIO-2914

### CODE REVIEW

File: shell/src/main/java/alluxio/shell/command/SetTtlCommand.java [hunk 1]
```diff
@@ -81,8 +85,9 @@ public CommandLine parseAndValidateArgs(String... args) {
   @Override
   public int run(CommandLine cl) throws AlluxioException, IOException {
     String[] args = cl.getArgs();
-    long ttlMs = Long.parseLong(CommonUtils.stripLeadingAndTrailingQuotes(args[1]));
-    Preconditions.checkArgument(ttlMs >= 0, "TTL value must be >= 0");
+    //long ttlMs = Long.parseLong(CommonUtils.stripLeadingAndTrailingQuotes(args[1]));
```
Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/SetTtlCommand.java]: remove line 88

File: tests/src/test/java/alluxio/shell/command/SetTtlCommandIntegrationTest.java [hunk 2]
```diff
@@ -97,11 +97,4 @@ public void setTtlWithInvalidOperationValue() throws Exception {
     Assert.assertEquals(-1,
         mFsShell.run("setTtl", "-action", "invalid", filePath, String.valueOf(ttl)));
   }
-
-  @Test
-  public void setTtlNegative() throws Exception {
```
Comment (apc999) [inline tests/src/test/java/alluxio/shell/command/SetTtlCommandIntegrationTest.java]: - why removing this?
- can we have a test where ttl input is with a unit?

Comment (maboelhassan) [inline tests/src/test/java/alluxio/shell/command/SetTtlCommandIntegrationTest.java]: Before my changes: negative time values were interpreted by apache commons cli as invalid command line options( because they start with "-" ). However, The exeuction of the command continues and the value is later checked if it is negative or not and if it is negative we throw another exception with the "TTL value must be >= 0" which is redundant given that the user doesn't even also see the "TTL value must be >= 0" message.

Now, I added the if condition in line 69 in setTtlCommand file such that if there is a negative value and apache cli fails to parse the command then the execution will stop at this point and the negative value will not have any further effect on anything else. That's why we no longer need to verify or unit test entering negative time values.

I added a unit test for entering values with different time units.

File: tests/src/test/java/alluxio/shell/command/SetTtlCommandIntegrationTest.java [hunk 3]
```diff
@@ -99,9 +99,27 @@ public void setTtlWithInvalidOperationValue() throws Exception {
   }
 
   @Test
-  public void setTtlNegative() throws Exception {
-    FileSystemTestUtils.createByteFile(mFileSystem, "/testFile", WriteType.MUST_CACHE, 1);
-    mFsShell.run("setTtl", "/testFile", "\"-1\"");
-    Assert.assertTrue(mOutput.toString().contains("TTL value must be >= 0"));
+  public void setTtlWithDifferentUnitTime() throws Exception {
```
Comment (uronce-cc) [inline tests/src/test/java/alluxio/shell/command/SetTtlCommandIntegrationTest.java]: add a test case for invalid time

File: tests/src/test/java/alluxio/shell/command/SetTtlCommandIntegrationTest.java [hunk 4]
```diff
@@ -99,9 +99,27 @@ public void setTtlWithInvalidOperationValue() throws Exception {
...
+    String filePath = "/testFile";
+    FileSystemTestUtils.createByteFile(mFileSystem, filePath, WriteType.MUST_CACHE, 1);
+    Assert.assertEquals(Constants.NO_TTL,
+        mFileSystem.getStatus(new AlluxioURI("/testFile")).getTtl());
+
+    AlluxioURI uri = new AlluxioURI("/testFile");
+    String[] timeUnits = {"", "ms", "millisecond", "s", "sec", "second", "m", "min", "minute", "h",
```
Comment (uronce-cc) [inline tests/src/test/java/alluxio/shell/command/SetTtlCommandIntegrationTest.java]: it's hard to maintain these two lists, we need to manually make sure the corresponding items have the same index...

what about using a `Map<String, long>`?

File: shell/src/main/java/alluxio/shell/command/SetTtlCommand.java [hunk 5]
```diff
@@ -92,7 +96,7 @@ public int run(CommandLine cl) throws AlluxioException, IOException {
 
   @Override
   public String getUsage() {
-    return "setTtl [-action delete|free] <path> <time to live(in milliseconds)>";
+    return "setTtl [-action delete|free] <path> <time to live>";
```
Comment (uronce-cc) [inline shell/src/main/java/alluxio/shell/command/SetTtlCommand.java]: document what format is `<time to live>`

File: tests/src/test/java/alluxio/shell/command/SetTtlCommandIntegrationTest.java (hunk 2 above)
Comment (apc999) [inline tests/src/test/java/alluxio/shell/command/SetTtlCommandIntegrationTest.java]: ack. thanks.

File: core/common/src/main/java/alluxio/exception/ExceptionMessage.java [hunk 6]
```diff
@@ -166,6 +166,7 @@
   DESTINATION_CANNOT_BE_FILE(
       "The destination cannot be an existing file when the source is a directory or a list of "
           + "files."),
+  VALUE_NOT_VALID_TIME("{0} is not valid time"),
```
Comment (uronce-cc) [inline core/common/src/main/java/alluxio/exception/ExceptionMessage.java]: `INVALID_TIME`

File: tests/src/test/java/alluxio/shell/command/SetTtlCommandIntegrationTest.java [hunk 7]
```diff
@@ -99,9 +102,43 @@ public void setTtlWithInvalidOperationValue() throws Exception {
...
   @Test
-  public void setTtlNegative() throws Exception {
+  public void setTtlWithDifferentUnitTime() throws Exception {
+    String filePath = "/testFile";
+    FileSystemTestUtils.createByteFile(mFileSystem, filePath, WriteType.MUST_CACHE, 1);
+    Assert.assertEquals(Constants.NO_TTL,
+        mFileSystem.getStatus(new AlluxioURI("/testFile")).getTtl());
+
+    AlluxioURI uri = new AlluxioURI("/testFile");
+    HashMap<String, Long> timeUnits = new HashMap<String, Long>();
```
Comment (maboelhassan) [inline tests/src/test/java/alluxio/shell/command/SetTtlCommandIntegrationTest.java]: Is there a more "elegant" way for initializing a map in Java? It seems that at some point in every approach we have to use the put method to add key value pairs.

Comment (apc999) [inline tests/src/test/java/alluxio/shell/command/SetTtlCommandIntegrationTest.java]: this is good :)

**Comment by maboelhassan** [comment]
Hey guys,
This pul request has been around for some time :) Can someone have a look please?
